src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Windows/Native/Helpers.cs

```diff
@@ -85,6 +85,7 @@ public static void DecodeObject(this byte[] encoded, CryptDecodeObjectStructType
                     throw Marshal.GetLastWin32Error().ToCryptographicException();
 
                 byte* decoded = stackalloc byte[cb];
+                new Span<byte>(decoded, cb).Clear();
```

Comment:
Is this actually needed for CryptDecodeObjectPointer here, or is this just a "defense in depth"?

Reply:
Strangely it was actually needed and I didn't invest heavily in figuring out why.

Reply:
Which case was it needed? (I am worried that there is a bug that this is masking.)

Reply:
> Which case was it needed? (I am worried that there is a bug that this is masking.)

This one, L88, for `decoded`.  Without this Clear call, this test fails:
```
      System.Security.Cryptography.X509Certificates.Tests.ExtensionsTests.KeyUsageExtension_BER [FAIL]
        Assert.Equal() Failure
        Expected: DigitalSignature
        Actual:   69402752
        Stack Trace:
          d:\repos\corefx\src\System.Security.Cryptography.X509Certificates\tests\ExtensionsTests.cs(199,0): at System.Security.Cryptography.X509Certificates.Tests.ExtensionsTests.KeyUsageExtension_BER()
```
(Of course, the other two Clear calls I added might also be working around some issue, but removing them doesn't cause any test failures.)

Reply:
@bartonjs Is it intentional that the `X509_KEY_USAGE` is set as 16-bit int, but read as 32-bit int?

https://github.com/dotnet/corefx/blob/a10890f4ffe0fadf090c922578ba0e606ebdd16c/src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Windows/X509Pal.CustomExtensions.cs#L30

vs.

https://github.com/dotnet/corefx/blob/a10890f4ffe0fadf090c922578ba0e606ebdd16c/src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Windows/X509Pal.CustomExtensions.cs#L54

We are reading a memory that is not getting set by the decode API. It just happens when we clear the memory ourselves upfront, everything works fine.

Reply:
Probably not.  Desktop introduced the asymmetry (https://referencesource.microsoft.com/#System/security/system/security/cryptography/x509/x509extension.cs,124), probably back in net20; but it copied into a local buffer.  During the port to core the code must have gotten "simplified" and since it worked it has been running since.

I don't know if the Windows implementation will always report cbData as 2 (making reading as a `ushort*` fine), or if they report 0/1 when the data is smaller.

Since DigitalSignature is in the low byte this could be easily validated with reading via `ushort*` and adding in a stopgap of span.Fill(0xFF) instead of span.Clear()

Reply:
> Since DigitalSignature is in the low byte this could be easily validated with reading via ushort* and adding in a stopgap of span.Fill(0xFF) instead of span.Clear()

Since this is pre-existing and since you're in the best position Jeremy to figure out the right fix here, I'm going to open an issue and go ahead with this change including the Clears, which can then hopefully be removed subsequently.